Mutual Auth Docs (#25509)
cilium-team/cilium: Documentation/network/servicemesh/mutual-authentication/mutual-authentication-example.rst

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _gs_mutual_authentication_example:

*****************************
Mutual Authentication Example
*****************************

This example shows you how to enforce mutual authentication between two Pods. 

Deploy a client (pod-worker) and a server (echo) using the following manifest:

.. code-block:: shell-session

    $ kubectl apply -f |SCM_WEB|\/examples/kubernetes/servicemesh/mutual-auth-example.yaml
    $ kubectl apply -f |SCM_WEB|\/examples/kubernetes/servicemesh/cnp-without-mutual-auth.yaml
    service/echo created
    deployment.apps/echo created
    pod/pod-worker created
    ciliumnetworkpolicy.cilium.io/simple-policy created 

Verify that the Pods have been successfully deployed:

.. code-block:: shell-session

    $ kubectl get svc echo
    NAME   TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)    AGE
    echo   ClusterIP   10.96.16.90   <none>        8080/TCP   42m
    $ kubectl get pod pod-worker 
    NAME         READY   STATUS    RESTARTS   AGE
    pod-worker   1/1     Running   0          40m

> Verify that the network policy has been deployed successfully and filters the traffic as expected. Run the following commands.

Reviewer: ```suggestion
Verify that the network policy has been deployed successfully and filters the traffic as expected. Run the following commands:
```

.. code-block:: shell-session

    $ kubectl exec -it pod-worker -- curl -s -o /dev/null -w "%{http_code}" http://echo:8080/headers
    200
    $ kubectl exec -it pod-worker -- curl http://echo:8080/headers-1
    Access denied

The first request should be successful (the *pod-worker* Pod is able to connect to the *echo* Service over a specific HTTP path and the HTTP status code is ``200``).
The second one should be denied (the *pod-worker* Pod is unable to connect to the *echo* Service over a specific HTTP path other than '/headers').

Before we enable mutual authentication between ``pod-worker`` and ``echo``, let's verify that the SPIRE server is healthy.

Assuming you have followed the installation instructions and have a SPIRE server serving Cilium, adding mutual authentication simply requires 
adding ``authentication.mode: "required"`` in the ingress/egress block in your network policies.

Verify SPIRE Health
===================

.. note::

    This example assumes a default SPIRE installation.

Let's first verify that the SPIRE server and agents automatically deployed are working as expected.

> Let's start with checking the SPIRE server and agents have been deployed.

Reviewer: These two lines are a bit redundant. Can we pick one or merge them together?

    Server is healthy.

Verify the list of attested agents:

.. code-block:: shell-session

    $ kubectl exec -n cilium-spire spire-server-0 -c spire-server -- /opt/spire/bin/spire-server agent list
    Found 2 attested agents:

    SPIFFE ID         : spiffe://spiffe.cilium/spire/agent/k8s_psat/default/64745bf2-bd9d-4e42-bb2b-e095a6b65121
    Attestation type  : k8s_psat
    Expiration time   : 2023-07-04 18:39:50 +0000 UTC
    Serial number     : 110848236251310359782141595494072495768

    SPIFFE ID         : spiffe://spiffe.cilium/spire/agent/k8s_psat/default/d4a8a6da-d808-4993-b67a-bed250bbc53e
    Attestation type  : k8s_psat
    Expiration time   : 2023-07-04 18:39:55 +0000 UTC
    Serial number     : 7806033782886940845084156064765627978

Notice that the SPIRE Server uses Kubernetes Projected Service Account Tokens (PSATs) to verify the identity of a SPIRE Agent running on a Kubernetes Cluster. 
Projected Service Account Tokens provide additional security guarantees over traditional Kubernetes Service Account Tokens and when supported by a Kubernetes cluster, PSAT is the recommended attestation strategy.

> Verify SPIFFE identity

Reviewer: Can we change "identity" to "Identities"? This section involves checking multiple identities and sections should use title case.

    SPIFFE ID        : spiffe://spiffe.cilium/cilium-agent
    Parent ID        : spiffe://spiffe.cilium/ns/cilium-spire/sa/spire-agent
    Revision         : 0
    X509-SVID TTL    : default
    JWT-SVID TTL     : default
    Selector         : k8s:ns:kube-system
    Selector         : k8s:sa:cilium

    Entry ID         : 8aa91d65-16c4-48a0-bc1f-c9bf26e6a25f
    SPIFFE ID        : spiffe://spiffe.cilium/cilium-operator
    Parent ID        : spiffe://spiffe.cilium/ns/cilium-spire/sa/spire-agent
    Revision         : 0
    X509-SVID TTL    : default
    JWT-SVID TTL     : default
    Selector         : k8s:ns:kube-system
    Selector         : k8s:sa:cilium-operator

Verify that the workloads have identities registered with the SPIRE server.

First, let's get the Cilium Identity of the *echo* Pod first as the SPIFFE ID - that uniquely identifies a workload - is partially based on the Cilium identity.

.. code-block:: shell-session

    $ IDENTITY_ID=$(kubectl get cep -l app=echo -o=jsonpath='{.items[0].status.identity.id}')
    $ echo $IDENTITY_ID
    17947

The SPIFFE ID for this workload is based on the ``spiffe://spiffe.cilium/identity/$IDENTITY_ID`` format.

> Verify that this Pod has a registered identity on the SPIRE server:

Reviewer: Can we work on rewording this part a bit to make it more clear? Normally, when the format of "Verify that...." is used, users will expect a command to execute. Since you are writing out multiple steps, let's add a transition statement. Additionally, the fact that the SPIFFE ID for a workload is based on the Cilium identity is mentioned in two separate places, which we can condense. For instance, we can do something like this:

```suggestion
Next, verify that the *echo* Pod has an identity registered with the SPIRE server.

To do this, you must first construct the Pod's SPIFFE ID. The SPIFFE ID for a workload is based on the ``spiffe://spiffe.cilium/identity/$IDENTITY_ID`` format, where `$IDENTITY_ID` is a workload's Cilium Identity.

Grab the Cilium Identity for the *echo* Pod;

.. code-block:: shell-session

    $ IDENTITY_ID=$(kubectl get cep -l app=echo -o=jsonpath='{.items[0].status.identity.id}')
    $ echo $IDENTITY_ID
    17947

Use the Cilium Identity for the *echo* pod to construct its SPIFFE ID and check it is registered on the SPIRE server:

.. code-block:: shell-session

>     # kubectl exec -n cilium-spire spire-server-0 -c spire-server -- /opt/spire/bin/spire-server entry show -spiffeID spiffe://spiffe.cilium/identity/$IDENTITY_ID

Reviewer 2: Some of your commands use `$` for a prompt (good), this one uses `#` which indicates it should be run as root (but is it the case?), some below don't have a prompt at all. I know we're not consistent with prompts in the docs in general, but it would be nice to be consistent at least in a new page we add. Could you please harmonise them (`#` for command that should be run as root, `$` everywhere else)?

Author: Don't think root is needed anywhere - corrected to `$` everywhere.

    Found 1 entry
    Entry ID         : 9fc13971-fb19-4814-b9f0-737b30e336c6
    SPIFFE ID        : spiffe://spiffe.cilium/identity/17947
    Parent ID        : spiffe://spiffe.cilium/cilium-operator
    Revision         : 0
    X509-SVID TTL    : default
    JWT-SVID TTL     : default
    Selector         : cilium:mutual-auth

> You can see the that the *cilium-operator* was listed in the ``Parent ID``. That is because the Cilium operator creates SPIRE entries for Cilium identity, as they are created.

Reviewer: ```suggestion
You can see the that the *cilium-operator* was listed in the ``Parent ID``. That is because the Cilium operator creates SPIRE entries for Cilium Identities as they are created.
```

> To get all registration entries, execute the following command:

Reviewer: Was this a typo?

```suggestion
To get all registered entries, execute the following command:
```

.. code-block:: shell-session

    kubectl exec -n cilium-spire spire-server-0 -c spire-server -- /opt/spire/bin/spire-server entry show -selector cilium:mutual-auth

There are as many entries as there are identities. Verify that these match by running the command:

.. code-block:: shell-session
    
    kubectl get ciliumidentities

The identify ID listed under ``NAME`` should match with the digits at the end of the SPIFFE ID executed in the previous command.

Enforce Mutual Authentication
=============================

> Rolling out mutual authentication with Cilium is as simple as adding the following to an existing or new CiliumNetworkPolicy.

Reviewer: ```suggestion
Rolling out mutual authentication with Cilium is as simple as adding the following to an existing or new CiliumNetworkPolicy:
```

> Add the following to your egress or ingress rules:

Reviewer: Can we merge this sentence with the one above? It is a bit redundant.

> .. code-block:: shell-session
> 
>     authentication:
>         mode: "required"

Reviewer 2: Isn't it YAML?

```suggestion
.. code-block:: yaml

    authentication:
        mode: "required"
```

Update the existing rule to only allow ingress access to mutually authenticated workloads to access *echo* using:

.. parsed-literal::

    $ kubectl apply -f |SCM_WEB|\/examples/kubernetes/servicemesh/cnp-with-mutual-auth.yaml

> Verify Mutual Authentication
> ===============================

Reviewer 2: Nit

```suggestion
Verify Mutual Authentication
============================
```

> Re-try your connectivity tests. They should give similar results as before. Has mutual authentication actually happened?

Reviewer: ```suggestion
Re-try your connectivity tests. They should give similar results as before:
```
We answer this question in the next step, so having it here may be confusing to users.

.. code-block:: shell-session

    $ kubectl exec -it pod-worker -- curl -s -o /dev/null -w "%{http_code}" http://echo:8080/headers
    200
    $ kubectl exec -it pod-worker -- curl http://echo:8080/headers-1
    Access denied

> Verify that mutual authentication has happened by accessing the logs on the agent. Start by enabling debug level (avoid doing this in production):

Reviewer: Can you add why should a user avoid enabling debug level in production? A warning block may be appropriate here.

Author: I've actually removed the "avoid doing this in production" as I've not seen any evidence we don't recommend it.

.. code-block:: shell-session

    cilium config set debug true

Examine the logs on the Cilium agent located in the same node as the *echo* Pod. 

> For brevity, we search for some specific log messages:

Reviewer: ```suggestion
For brevity, you can search for some specific log messages:
```

.. code-block:: shell-session

    $ kubectl -n kube-system -c cilium-agent logs cilium-9pshw --timestamps=true | grep "Policy is requiring authentication\|Validating Server SNI\|Validated certificate\|Successfully authenticated"
    2023-07-04T17:58:28.795760597Z level=debug msg="Policy is requiring authentication" auth_type=spire local_identity=17947 remote_identity=39239 subsys=auth
    2023-07-04T17:58:28.800509503Z level=debug msg="Validating Server SNI" SNI ID=39239 subsys=auth
    2023-07-04T17:58:28.800525190Z level=debug msg="Validated certificate" subsys=auth uri-san="[spiffe://spiffe.cilium/identity/39239]"
    2023-07-04T17:58:28.801441968Z level=debug msg="Successfully authenticated" auth_type=spire local_identity=17947 remote_identity=39239 remote_node_ip=10.0.1.175 subsys=auth

> When you apply a mutual authentication policy, the agent retrieves the identity of the source Pod, connects to the node where the destination Pod is running and performs a TLS handshake.

Reviewer: As someone reading this who isn't as familiar with mTLS, this contradicts what you say in "Mutual Authentication and mTLS Background" because a TLS handshake is a one-way authentication, not necessarily a mutual authentication. I understand that in the logs here it is more clear that this agent is validating a remote agent, but I think it would make sense to mention that the other agent will also authenticate this agent.

> As the TLS handshake succeeded, then the connection was authenticated, the traffic protected by policy could proceed and packets between the two Pods can flow until the network policy is removed or the entry expires.

Reviewer: Let's break this up in to two sentences:

```suggestion
As the TLS handshake succeeded, the connection was authenticated and the traffic protected by policy could proceed. Packets between the two Pods can flow until the network policy is removed or the entry expires.
```

Can you expand on what you mean by an entry expiring here? When does an entry expire?
